Source/WebCore/style/RuleSet.cpp

```diff
@@ -160,6 +160,28 @@ void RuleSet::addRule(RuleData&& ruleData, CascadeLayerIdentifier cascadeLayerId
     storeIdentifier(scopeRuleIdentifier, m_scopeRuleIdentifierForRulePosition);
 
     const auto& scopeRules = scopeRulesFor(ruleData);
+
+    auto computeLinkMatchType = [&] {
+        // General case: no @scope rule or current rule selector is not :scope.
+        if (scopeRules.isEmpty() || !ruleData.selector()->hasScope())
+            return SelectorChecker::determineLinkMatchType(ruleData.selector());
+
+        // When current rule is :scope, we need to take into account the @scope selectors to determine the link match type.
+        auto scopingRootLinkMatchType = [&] -> unsigned {
+            auto scopeRule = scopeRules.last();
+            if (!scopeRule->scopeStart().isEmpty()) {
+                // FIXME: Support multiple selectors
+                auto scopingRootLinkMatchType = SelectorChecker::determineLinkMatchType(scopeRule->scopeStart().first());
+                if (!scopeRule->scopeEnd().isEmpty())
+                    scopingRootLinkMatchType &= SelectorChecker::determineLinkMatchType(scopeRule->scopeEnd().first());
+                return scopingRootLinkMatchType;
+            }
+            return SelectorChecker::MatchAll;
+        };
+        return SelectorChecker::determineLinkMatchType(ruleData.selector(), scopingRootLinkMatchType());
+    };
+    ruleData.setLinkMatchType(computeLinkMatchType());
```

Comment:
Some more of the stuff above could probably still move to SelectorChecker::determineLinkMatchType (by passing the scopeRule)